Federated Model (seq-to-seq)/model_utils.py

```diff
@@ -0,0 +1,414 @@
+        slot_accuracy = tf.math.equal(slot_accuracy,
+                                      tf.math.reduce_sum(slot_mask, axis=-1))
+
+        intent_tp = tf.math.equal(
+            y_true[1][:, 0],
+            tf.math.argmax(y_pred[1], axis=-1, output_type=tf.int32))
+        tp = tf.reduce_sum(
+            tf.cast(tf.math.logical_and(slot_accuracy, intent_tp), tf.float32))
+
+        total = tf.reduce_sum(tf.ones_like(y_true[1], dtype=tf.float32))
+
+        self.true_positives.assign_add(tp)
+        self.total.assign_add(total)
+
+    def result(self):
+        return tf.math.divide(self.true_positives, self.total)
+
+    def reset_states(self):
+        self.true_positives.assign(0)
+        self.total.assign(0)
+
+
+# compute f1 score is modified from conlleval.pl
+def start_of_chunk(prev_tag, tag, prev_tag_type, tag_type, chunk_start=False):
+    """Checks if the current tag indicates the start of a new chunk.
+
+    Args:
+    prev_tag: The tag of the previous token.
+    tag: The tag of the current token.
+    prev_tag_type: The tag type of the previous token.
+    tag_type: The tag type of the previous token.
+
+    Returns:
+    A boolean variable indicating if the current tag is the beginning of a new
+    chunk.
+    """
+
+    if prev_tag == 'B' and tag == 'B':
+        chunk_start = True
+    if prev_tag == 'I' and tag == 'B':
+        chunk_start = True
+    if prev_tag == 'O' and tag == 'B':
+        chunk_start = True
+    if prev_tag == 'O' and tag == 'I':
+        chunk_start = True
+
+    if tag != 'O' and tag != '.' and prev_tag_type != tag_type:
+        chunk_start = True
+    return chunk_start
+
+
+def end_of_chunk(prev_tag, tag, prev_tag_type, tag_type, chunk_end=False):
+    """Checks if the current tag indicates the end of a chunk.
+
+    Args:
+    prev_tag: The tag of the previous token.
+    tag: The tag of the current token.
+    prev_tag_type: The tag type of the previous token.
+    tag_type: The tag type of the previous token.
+
+    Returns:
+    A boolean variable indicating if the current tag is the end of a
+    chunk.
+    """
+
+    if prev_tag == 'B' and tag == 'B':
+        chunk_end = True
+    if prev_tag == 'B' and tag == 'O':
+        chunk_end = True
+    if prev_tag == 'I' and tag == 'B':
+        chunk_end = True
+    if prev_tag == 'I' and tag == 'O':
+        chunk_end = True
+
+    if prev_tag != 'O' and prev_tag != '.' and prev_tag_type != tag_type:
+        chunk_end = True
+    return chunk_end
+
+
+def split_tag_type(tag):
+    """Extracts and returns the tag and tag type.
+    """
+
+    split_tag = tag.split('-')
+    if len(split_tag) > 2 or len(split_tag) == 0:
+        raise ValueError('tag format wrong. it must be B-xxx.xxx')
+    if len(split_tag) == 1:
+        tag = split_tag[0]
+        tag_type = ""
+    else:
+        tag = split_tag[0]
+        tag_type = split_tag[1]
+    return tag, tag_type
+
+
+def compute_f1(correct_slots, pred_slots):
```

Comment:
This function seems to be a bit long and hard to follow, could you break it into smaller pieces?